src/libsync/pushnotifications.cpp

```diff
@@ -165,33 +175,128 @@ void PushNotifications::handleAuthenticated()
     qCInfo(lcPushNotifications) << "Authenticated successful on websocket";
     _failedAuthenticationAttemptsCount = 0;
     _isReady = true;
+    startPingTimeoutTimer();
     emit ready();
+
+    // We maybe reconnected to websocket while being offline for a
+    // while. To not miss any notifications that may have happend,
+    // emit all the signals once.
+    emitFilesChanged();
+    emitNotificationsChanged();
+    emitActivitiesChanged();
 }
 
 void PushNotifications::handleNotifyFile()
 {
     qCInfo(lcPushNotifications) << "Files push notification arrived";
-    emit filesChanged(_account);
+    emitFilesChanged();
 }
 
 void PushNotifications::handleInvalidCredentials()
 {
     qCInfo(lcPushNotifications) << "Invalid credentials submitted to websocket";
     if (!tryReconnectToWebSocket()) {
-        _isReady = false;
+        closeWebSocket();
         emit authenticationFailed();
     }
 }
 
 void PushNotifications::handleNotifyNotification()
 {
     qCInfo(lcPushNotifications) << "Push notification arrived";
-    emit notificationsChanged(_account);
+    emitNotificationsChanged();
 }
 
 void PushNotifications::handleNotifyActivity()
 {
     qCInfo(lcPushNotifications) << "Push activity arrived";
+    emitActivitiesChanged();
+}
+
+void PushNotifications::onWebSocketPongReceived(quint64 /*elapsedTime*/, const QByteArray &payload)
+{
+    handleTimeoutPong(payload);
+}
+
+void PushNotifications::handleTimeoutPong(const QByteArray &payload)
+{
+    // We are not interested in different pongs from the server
+    const auto expectedPingPayload = timeoutPingPayload();
+    if (payload != expectedPingPayload) {
+        return;
+    }
+
+    qCDebug(lcPushNotifications) << "Pong received in time";
+
+
+    _timeoutPongReceivedFromWebSocketServer = true;
+    startPingTimeoutTimer();
+}
+
+void PushNotifications::startPingTimeoutTimer()
+{
+    _pingTimeoutTimer.stop();
+    _pingTimer.setInterval(_pingTimeoutInterval);
+    _pingTimer.setSingleShot(true);
+    connect(&_pingTimer, &QTimer::timeout, this, &PushNotifications::pingWebSocketServer, Qt::UniqueConnection);
+    _pingTimer.start();
+}
+
+void PushNotifications::startPingTimedOutTimer()
+{
+    _pingTimeoutTimer.setInterval(_pingTimeoutInterval);
+    _pingTimeoutTimer.setSingleShot(true);
+    connect(&_pingTimeoutTimer, &QTimer::timeout, this, &PushNotifications::onPingTimedOut, Qt::UniqueConnection);
+    _pingTimeoutTimer.start();
+}
+
+QByteArray PushNotifications::timeoutPingPayload() const
+{
+    const void *push_notifications_address = this;
```

Comment:
@FlexW Isn't this too optimistic? I mean, what if the address of **this** changes? Maybe it makes sense to use something more constant for this? Like, user name? **user@nextcloud.com**?

Reply:
Why should this **this** change? If **this** changes then that means that the PushNotifications object was recreated. In this case the pong handler will also be gone and I#m not interested in that pong anymore. I would even say the address of the object is better than some string because I'm clearly just interested in a pong that got sent in response of a ping from exactly this object. Additionally it costs a lot less bandwith then sending some string. I think we could even go without a payload and accept everything as fine as long as we get any pongs.